tools/buildtools/schc.py

```diff
@@ -0,0 +1,923 @@
+    TokenType.IF_BEFORETIME:        TokenProperties(isConditionalBranch=True,   hasArguments=True, isGeneric=True, needsToInvert=True, shortVersion=TokenType.IF_BEFORETIME_S, longVersion=TokenType.IF_BEFORETIME_L),
+    TokenType.IF_SINCETIME:         TokenProperties(isConditionalBranch=True,   hasArguments=True, isGeneric=True,                     shortVersion=TokenType.IF_SINCETIME_S, longVersion=TokenType.IF_SINCETIME_L),
+    TokenType.BRANCH:               TokenProperties(isUnconditionalBranch=True, hasArguments=True, isGeneric=True,                     shortVersion=TokenType.BRANCH_S, longVersion=TokenType.BRANCH_L),
+
+    # Extra tokens
+    TokenType.ELSE:                 TokenProperties(isExtraToken=True),
+    TokenType.BRACE_OPEN:           TokenProperties(isExtraToken=True),
+    TokenType.BRACE_CLOSE:          TokenProperties(isExtraToken=True),
+    TokenType.ARGS:                 TokenProperties(isExtraToken=True),
+    TokenType.NOT:                  TokenProperties(isExtraToken=True),
+    TokenType.LABEL:                TokenProperties(isExtraToken=True),
+}
+
+tokenLiterals: dict[str, TokenType] = {x.value: x for x in TokenType}
+
+
+regex_label = re.compile(r"(?P<label>\w+)\s*:")
+regex_identifier = re.compile(r"(?P<identifier>\w+)")
+regex_individualTokens = re.compile(r"(?P<individual>[\{\}])")
+
+@dataclasses.dataclass
+class Token:
+    tokenType: TokenType
+
+    # The literal read from the input. Some token types don't have a fixed literal, like LABEL or ARGS
+    tokenLiteral: str
+
+    # Track the token position on the original file, for better error messages
+    filename: str
+    lineNumber: int
+    columnNumber: int
+
+    def getProperties(self) -> TokenProperties:
+        return tokenPropertiesDict[self.tokenType]
+
+    def newFromTokenType(self, newType: TokenType) -> Token:
+        return Token(newType, newType.value, self.filename, self.lineNumber, self.columnNumber)
+
+    def newFromTokenTypePreserveLiteral(self, newType: TokenType) -> Token:
+        return Token(newType, self.tokenLiteral, self.filename, self.lineNumber, self.columnNumber)
+
+    def __str__(self) -> str:
+        ret = f"{self.filename}:{self.lineNumber}:{self.columnNumber}\n"
+        ret += f"  {self.tokenType.name}"
+        if self.tokenLiteral != self.tokenType.value:
+            ret += f" '{self.tokenLiteral}'"
+        return ret
+
+
+# Allows to know which token will be the next one, even with recursive functions
+class TokenIterator:
+    def __init__(self, tokens: list[Token]):
+        self.tokens = list(tokens)
+        self.index = 0
+
+    def get(self) -> Token|None:
+        if self.index >= len(self.tokens):
+            return None
+        token = self.tokens[self.index]
+        self.index += 1
+        return token
+
+    def unget(self) -> None:
+        self.index -= 1
+        if self.index < 0:
+            raise RuntimeError()
```

Comment:
```suggestion
            raise Exception("bad programming, can't unget")
```
:shrug: on the message but I do think you should have one, and also use Exception, or make a `class SchcError(Exception): pass`

Also I think the check should come before decrementing index so you don't leave the class in a bad state